kernel/posix/semaphore.c

```diff
@@ -136,3 +139,342 @@ int sem_wait(sem_t *semaphore)
+		return 0;
+	}
+
+	it = nsem_otable;
+	lmt = nsem_otable + NSEM_SYS_OPEN_MAX;
+
+	while (it != lmt) {
+		if (it->nsem_ptr == nsem) {
+			++count;
+		}
+		++it;
+	}
+
+	return count;
+}
+
+static inline unsigned nsem_otable_count_opened(k_tid_t proc)
+{
+	struct _nsem_otable *it, *lmt;
+	unsigned int count = 0;
+
+	it = nsem_otable;
+	lmt = nsem_otable + NSEM_SYS_OPEN_MAX;
+
+	while (it != lmt) {
+		if (it->proc == proc) {
+			++count;
+		}
+		++it;
+	}
+
+	return count;
+}
+
+/*
+ * Flags control if the semaphore is created or simply retrieved
+ * O_CREAT: Creates a new semaphore with mode_t permissions and count.
+ * O_EXCL:  Passed along with O_CREAT, fails if a semaphore with name already
+ *          exists.
+ *
+ * This function is atomic since a few variables are shared.
+ */
+sem_t *sem_open(const char *name, int oflag, ...)
+{
+	struct _nsem *nsem;
+	unsigned int value;
+	unsigned int key;
+	va_list al;
+	k_tid_t curr;
+
+	key = irq_lock();
+
+	nsem = nsem_find_by_name(name);
+	curr = k_current_get();
+
+	/*
+	 * Multiple calls to sem_open() by the same process
+	 * should be successful.
+	 */
+	if (nsem_otable_find(curr, nsem)) { /* Deals with NULL */
+		goto success;
+	}
+
+	if (sem_named_n == NSEM_SYS_N_MAX) {
+		SEM_ERRNO_SET(ENFILE);
+	}
+
+	if (nsem_otable_count_opened(curr) == NSEM_PROC_OPEN_MAX) {
+		SEM_ERRNO_SET(EMFILE);
+	}
+
+	/* Check if this named semaphore already exists */
+	if (nsem) {
+		if (oflag & O_EXCL) {
+			SEM_ERRNO_SET(EEXIST);
+		}
+	} else {
+		if (!(oflag & O_CREAT)) {
+			SEM_ERRNO_SET(ENOENT);
+		}
+
+		va_start(al, oflag);
+		va_arg(al, mode_t); /* permissions not used */
+		value = va_arg(al, unsigned int);
+		va_end(al);
+
+		nsem = nsem_alloc(name);
+		sem_init(&nsem->sem, 0, value);
+		sem_named_n++;
+	}
+
+	/* Should always be successful, we checked the size before */
+	nsem_otable_alloc(curr, nsem);
+
+success:
+	irq_unlock(key);
+	return &nsem->sem;
+
+error:
+	irq_unlock(key);
+	return SEM_FAILED;
+}
+
+/*
+ * Close can destroy a semaphore too if sem_unlink has been called on that sem
+ * before and it's the last reference to it.
+ */
+int sem_close(sem_t *sem)
+{
+	struct _nsem_otable *ote;
+	struct _nsem *nsem;
+	unsigned int key;
+	k_tid_t curr;
+
+	key = irq_lock();
+
+
+	/* This is safe since we don't read or write nsem */
+	nsem = nsem_container_of(sem);
+	curr = k_current_get();
+	ote = nsem_otable_find(curr, nsem);
+
+	/* If ote is set, nsem is legit so we can use it */
+	if (!ote) {
+		SEM_ERRNO_SET(EINVAL);
+	}
+
+	/* Dereference semaphore */
+	nsem_otable_free(ote);
+	if (nsem->to_free && !nsem_otable_count_refs(nsem)) {
+		nsem_free(nsem);
+		sem_named_n--;
+	}
+
+	irq_unlock(key);
+	return 0;
+
+error:
+	irq_unlock(key);
+	return -1;
+}
+
+/*
+ * Free a semaphore if no references left.
+ * Otherwise, mark it as pending to free.
+ */
+int sem_unlink(const char *name)
+{
+	struct _nsem *nsem;
+	unsigned int key, refs;
+
+	key = irq_lock();
+	nsem = nsem_find_by_name(name);
```

Comment:
Do the find_by_name() outside the lock.  That doesn't need to be synchronized.

Reply:
Noted.